test/cases/testdata/strings/test-strings-0924.yaml

```diff
@@ -0,0 +1,56 @@
+cases:
```

Comment:
These tests are good, but it would be good to see a case or two with unicode characters. Should work given your implementation :)

Reply:
I added two unicode test cases. One that is quite straight forward I would say, and another that might have unexpected results (strings/reverse_unicode_multi_char_emojii). This is when a unicode glyph(?) consists of multiple codepoints, which is often the case for modern emojies. As the functionality is now, we reverse at the level of codepoints, which changes how the emojies are rendered.

Is the current functionality as expected? Or would we want to reorder at the glyph level instead?

Reply:
Some good research there! I'd say reordering at glyph level would be less surprising, or what do you think? It would obviously be (happily!) surprising if we saw policies relying on the reverse implementation of emojis 😆 but is this not case with e.g. characters like the German ß as well?

Reply:
I do agree that that would be the least surprising. I think the german ß is a single unicode code point (U+00DF), but some characters with accents etc. are examples of characters that are modified by following code points (there are some examples in the repository linked to below). 

After doing some research on the matter yesterday, I do believe that that this is more complicated than what I would have guessed beforehand. I'd like to present some findings on the matter to spark some discussion:

- A visual character as interpreted by a human is called a grapheme cluster (not glyph as I first suggested) in unicode therms. What we want to accomplish is to reverse at the level of grapheme. To cite from https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries:
  > It is important to recognize that what the user thinks of as a “character" [...] this is called a user-perceived character. [...] These user-perceived characters are approximated by what is called a grapheme cluster, which can be determined programmatically
- I have not yet read said document in full, but I believe it details how grapheme-cluster boundaries can be determined. I've also found an example of a public go-library which does this: https://github.com/rivo/uniseg. I am not suggesting that this is included as a dependency, but rather want to provide it as an example.
- Implementing the strings.reverse this way would (arguably) be inconsistent with current implementations of string operations, like for instance split. Take a look at this example policy: https://play.openpolicyagent.org/p/aYjrhAiKJu. Is this also non-surprising?
- If we implement some way of doing grapheme-clustering in the strings.reverse function, both in go and wasm, would this consequently lead to continuous maintanence need of said functionality as unicode is expanded with new emojies and characters in the future? 

To summarise based on what I've written above: Unicode handling is hard, and in some cases depending on the use-case. 
What would be the best way forward on this matter?

Reply:
Oh wow, yeah the flags look like they're always composed of multiple code points. Funny thing that when you try out a policy with those included in the Playground.. on Firefox you can actually see the cursor moving halfway inside of the flag when pressing the right arrow, and you'll need to press twice to make it to the next one 😄 

Looks like we've pretty much just been lucky in previous unicode and character position tests like this one https://github.com/open-policy-agent/opa/blob/main/test/cases/testdata/strings/test-strings-indexof-unicode.yaml

Let's ignore these for now. As long as any "normal" characters (i.e. those found in common languages and likely to be prevalent in policy) works as expected, I think that's good enough. I'm sure there will be a trendy new startup somewhere creating policies exlusively using emojis, but if that's so I don't think it would be unfair to ask them for help with supporting their use case :)

Reply:
That's interesting, and even shows that established software like Firefox can have "surprises" when it comes to unicode handling.

Yep, for instance if one of the preceding emojis in "strings/indexof: unicode string emoji" were substituted with a flag, the want_result would have to be incremented by 1.

Hehe, that's a fair point.
Just to summarise what we can expect: Ascii should work without any surprises, and most latin-based characters I think, though there will be some exceptions like "ä" vs "ä". They look identical (on my computer at least), but the first is one code point, whereas the latter is two. Here is another example from the playground (using split, but the inner workings is the same as for strings.reverse): https://play.openpolicyagent.org/p/rvzH7nbNQW 
If "äe" (latter) would be reversed, I guess it will become "ëa", but if "äe" (first) is reversed, it becomes "eä". But as you say, characters such as these are probably not expected to be prevalent in policies.

Reply:
Agreed, let's make a best effort (which you've definitely done already!) and improve on that later if requested.